releases/release-5.1.2.md

```diff
@@ -0,0 +1,156 @@
+    - retry when meeting stablish conn fails [#2570](https://github.com/pingcap/tics/pull/2570)
+    - Support pushing down MOD(N, M) to TiFlash. [#2560](https://github.com/pingcap/tics/pull/2560)
+    - Add Grafana panels for write throughput per instance [#2544](https://github.com/pingcap/tics/pull/2544)
+    -  Cherry-pick #2216 #2460 #2494 & Optimizr read-index retry to release-5.1 [#2345](https://github.com/pingcap/tics/pull/2345)
+
+## Bug Fixes
+
++ TiDB
+
+    - Fix wrong index hash join when hash col is enum. [#28081](https://github.com/pingcap/tidb/pull/28081)
+    - Fix a batch client bug that recycle idle connection may block sending requests in some rare cases. [#27678](https://github.com/pingcap/tidb/pull/27678)
+    - Keep the overflow check logic same with mysql [#26725](https://github.com/pingcap/tidb/pull/26725)
+    - Fix #26147 that `tidb` returns an `unknow` error with no message while it should return the error that contains `pd is timeout`. [#26682](https://github.com/pingcap/tidb/pull/26682)
+    - Fix wrong charset and collation for case when function [#26673](https://github.com/pingcap/tidb/pull/26673)
+    - Fix the issue that greatest(datetime) union null returns empty string [#26566](https://github.com/pingcap/tidb/pull/26566)
+
++ TiFlash
+
+    - Fix the potential issue of data inconsistency after crashes when deployed on multi-disks [#2776](https://github.com/pingcap/tics/pull/2776)
+    - fix bug that SharedQueryBlockInputStream may loss block randomly. [#2762](https://github.com/pingcap/tics/pull/2762)
+    - Please add a release note, or a 'None' if it is not needed. [#2710](https://github.com/pingcap/tics/pull/2710)
+    - Fix concurrency between init store and DDL. [#2693](https://github.com/pingcap/tics/pull/2693)
+    - Fix false alert of expected internal DDL error [#2673](https://github.com/pingcap/tics/pull/2673)
+    - Fix the bug that filters unexpected data if the query contains "constant" "<"/"<="/">"/">=" "column" [#2635](https://github.com/pingcap/tics/pull/2635)
+    - Fix an overflow bug in decimal MOD(N, M). [#2609](https://github.com/pingcap/tics/pull/2609)
+    - Fix the potential panic when applying Snapshot with multiple DDL operations [#2606](https://github.com/pingcap/tics/pull/2606)
+    - Fix inaccurate store size on tiflash metric under heavy write scenario [#2598](https://github.com/pingcap/tics/pull/2598)
+    - Fix the potential issue that TiFlash cannot GC the delta data after running for a long time [#2337](https://github.com/pingcap/tics/pull/2337)
+
++ PD
+
+    - Fix the bug that PD would not fix down-peer in time. [#4083](https://github.com/tikv/pd/pull/4083)
+    - Fix an issue where data is not stored when using max-replicas or location-labels to indirectly update default placement rule [#3915](https://github.com/tikv/pd/pull/3915)
+    - Fix the bug that PD may panic during scaling out TiKV. [#3911](https://github.com/tikv/pd/pull/3911)
+
```

Comment:
Fix the bug that hot region scheduler can not work when the cluster has evict leader scheduler. [#3697](https://github.com/tikv/pd/pull/3697)